src/sinks/gcp/stackdriver_metrics.rs

```diff
@@ -0,0 +1,215 @@
+    pub request: TowerRequestConfig,
+    #[serde(default)]
+    pub batch: BatchConfig,
+    pub tls: Option<TlsOptions>,
+}
+
+impl_generate_config_from_default!(StackdriverConfig);
+
+inventory::submit! {
+    SinkDescription::new::<StackdriverConfig>("gcp_stackdriver_metrics")
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "gcp_stackdriver_metrics")]
+impl SinkConfig for StackdriverConfig {
+    async fn build(&self, cx: SinkContext) -> crate::Result<(VectorSink, Healthcheck)> {
+        let mut token = gouth::Builder::new().scopes(&[
+            "https://www.googleapis.com/auth/cloud-platform",
+            "https://www.googleapis.com/auth/monitoring",
+            "https://www.googleapis.com/auth/monitoring.write",
+        ]);
+
+        if let Some(credentials_path) = self.credentials_path.clone() {
+            token = token.file(credentials_path);
+        }
+
+        let token = token.build()?;
+        let healthcheck = healthcheck().boxed();
+        let started = chrono::Utc::now();
+        let request = self.request.unwrap_with(&REQUEST_DEFAULTS);
+        let tls_settings = TlsSettings::from_options(&self.tls)?;
+        let client = HttpClient::new(tls_settings)?;
+        let batch = BatchSettings::default()
+            .events(1)
+            .parse_config(self.batch)?;
+
+        let sink = HttpEventSink {
+            config: self.clone(),
+            started,
+            token,
+        };
+
+        let sink = BatchedHttpSink::new(
+            sink,
+            MetricsBuffer::new(batch.size),
+            request,
+            batch.timeout,
+            client,
+            cx.acker(),
+        )
+        .sink_map_err(
+            |error| error!(message = "Fatal gcp_stackdriver_metrics sink error.", %error),
+        );
+
+        Ok((VectorSink::Sink(Box::new(sink)), healthcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "gcp_stackdriver_metrics"
+    }
+}
+
+struct HttpEventSink {
+    config: StackdriverConfig,
+    started: DateTime<Utc>,
+    token: gouth::Token,
+}
+
+#[async_trait::async_trait]
+impl HttpSink for HttpEventSink {
+    type Input = Event;
+    type Output = Vec<Metric>;
+
+    fn encode_event(&self, event: Event) -> Option<Self::Input> {
+        let metric = event.into_metric();
+
+        match &metric.data.value {
+            &MetricValue::Counter { .. } => Some(Event::Metric(metric)),
+            &MetricValue::Gauge { .. } => Some(Event::Metric(metric)),
+            not_supported => {
+                warn!("Unsupported metric type: {:?}.", not_supported);
+                None
+            }
+        }
+    }
+
+    async fn build_request(
+        &self,
+        mut metrics: Self::Output,
+    ) -> crate::Result<hyper::Request<Vec<u8>>> {
+        let metric = metrics.pop().expect("only one metric");
+        let namespace = metric.namespace().unwrap_or("namespace");
```

Comment:
Is the literal `namespace` valid here?

Reply:
Yes, in the sense that we have to provide a value if the metric doesn't have one. It is used to property set up a custom metric type.

Reply:
Right, I mean is it valid to GCP? That is, does it make sense to default to this fixed string, or should it be a configuration item?

Reply:
I honestly don't know the right answer, I see value in supporting a fallback value if a metric doesn't have a namespace. GCP doesn't like metrics with the same type but having a different kind (Gauge vs Cumulative for instance). Defaulting the namespace value with `namespace` doesn't address that issue though.

Reply:
I introduced a `default_metric_namespace` property in the sink's configuration.